src/cache/gcs.rs

```diff
@@ -0,0 +1,353 @@
+        let metadata = fs::metadata(&self.credentials_path).chain_err(|| {
+            "Couldn't stat GCS credentials file"
+        })?;
+        if !metadata.is_file() {
+            bail!("Couldn't open GCS credentials file.");
+        }
+        let mut file = File::open(&self.credentials_path)?;
+        let mut service_account_json = String::new();
+        file.read_to_string(&mut service_account_json)?;
+        let sa_key: ServiceAccountKey = serde_json::from_str(&service_account_json)?;
+
+        let scope = (if self.read_only {
+            "https://www.googleapis.com/auth/devstorage.readonly"
+        } else {
+            "https://www.googleapis.com/auth/devstorage.read_write"
+        }).to_owned();
+
+        let jwt_claims = JwtClaims {
+            issuer: sa_key.client_email,
+            scope: scope,
+            audience: "https://www.googleapis.com/oauth2/v4/token".to_owned(),
+            expiration: expire_at.timestamp(),
+            issued_at: chrono::UTC::now().timestamp(),
+        };
+
+        let binary_key = openssl::rsa::Rsa::private_key_from_pem(
+            sa_key.private_key.as_bytes()
+        )?.private_key_to_der()?;
+
+        let auth_request_jwt = jwt::encode(
+            &jwt::Header::new(jwt::Algorithm::RS256),
+            &jwt_claims,
+            &binary_key,
+        )?;
+
+        Ok(auth_request_jwt)
+    }
+
+    pub fn credentials(&self, client: &Client<HttpsConnector>) -> SFuture<GCSCredential> {
+        let mut future_opt = self.cached_credentials.borrow_mut();
+
+        let needs_refresh = match Option::as_mut(&mut future_opt).map(|mut f| f.poll()) {
+            None => true,
+            Some(Ok(Async::Ready(ref creds))) => creds.expiration_time < chrono::UTC::now(),
+            _ => false
+        };
+
+        if needs_refresh {
+            let client = client.clone();
+            let expires_at = chrono::UTC::now() + chrono::Duration::minutes(59);
+            let auth_jwt = self.auth_request_jwt(&expires_at);
+            let credentials: SFuture<_> = Box::new(future::result(auth_jwt).and_then(move |auth_jwt| {
+                let url = "https://www.googleapis.com/oauth2/v4/token";
+                let params = form_urlencoded::Serializer::new(String::new())
+                    .append_pair("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer")
+                    .append_pair("assertion", &auth_jwt)
+                    .finish();
+
+                let mut request = Request::new(Method::Post, url.parse().unwrap());
+                {
+                    let mut headers = request.headers_mut();
+                    headers.set(ContentType("application/x-www-form-urlencoded".parse().unwrap()));
+                    headers.set(ContentLength(params.len() as u64));
+                }
+                request.set_body(params);
+
+                client.request(request).map_err(Into::into)
```

Comment:
Unfortunately, you have to manually map the error because otherwise the error types don't match-- `IntoFuture` has error as an associated type, so a given future can only implement `IntoFuture` for one error type. `hyper::FutureResponse` has its error type as `hyper::error::Error`, so we need to coerce it to our manual error type before we can return it.